### PR TITLE
front: add e2e stdcm tests

### LINKED ISSUE #9265: front: add e2e tests STDCM
## Description and goal

Steps:
1. Modify `small_infra` to add other OPs with different CHs (to be able to test CH selection)
2. Create a scenario with `small_infra` and trains to fill the grid
3. Send the infra id, the search time window and the timetable id of the scenario created to the endpoint `POST/stdcm/search_environment`
4. Run the tests (cf CA below)
## Acceptance criteria

- [ ] All the steps from the [manual test](https://github.com/osrd-project/osrd-tests/issues/76) are tested, but on `small_infra`. Up to step 29, skipping step 28 (the simulation sheet will be tested in another test).
- [ ] Dates and times are to be filled in by hand, but not the tolerance cause there's a risk of this changing, so we'll leave it for now.
- [ ] The map is tested on Chrome

### PR DESCRIPTION
**Pull request by Maymanaf**
Closes #9265

### CODE REVIEW

**Comment by Maymanaf** [comment]
The behavior in CI differs from the local machine (tested with WSL and Ubuntu); the simulation results aren’t the same. I'm currently investigating

**Comment by Yohh** [comment]
> The behavior in CI differs from the local machine (tested with WSL and Ubuntu); the simulation results aren’t the same. I'm currently investigating

you should wait for https://github.com/OpenRailAssociation/osrd/pull/9554 to be merged then rebase, it will change some behaviours for sure

**Comment by Khoyo** [comment]
#9554 shouldn't change any existing behavior, it only adds a script and a *new* docker image (the ci and running e2e tests outside of docker are completely unaffected)

**Comment by Yohh** [comment]
> #9554 shouldn't change any existing behavior, it only adds a script and a _new_ docker image (the ci and running e2e tests outside of docker are completely unaffected)

sure, I wasn't saying it about behaviors in the tests themselves

File: front/tests/pages/stdcm-page-model.ts
Comment (SharglutDev) [inline front/tests/pages/stdcm-page-model.ts]: image

It doesn't break the test but if you click again on the tolerance input, it reopens it, not close it. We shoudl click outside of the modal to close it but not on the same input again.

There is an [issue](https://github.com/OpenRailAssociation/osrd-ui/issues/693) to add a close button to date/time/tolerance pickers, it's planned for this PI.

If you want and if your PR passes before this issue, we can edit the acceptance criteria to modify the e2e test.

File: front/src/applications/stdcm/components/StdcmResults/StdcmResultsTable.tsx
Comment (Akctarus) [inline front/src/applications/stdcm/components/StdcmResults/StdcmResultsTable.tsx]:
```suggestion
            data-testid="all-vias-button"
```

File: front/tests/pages/stdcm-page-model.ts
Comment (SharglutDev) [inline front/tests/pages/stdcm-page-model.ts:490]: Seems like this does the same as the `downloadSimulation` function first lines. Maybe it's not necessary ?

File: front/tests/006-stdcm.spec.ts
Comment (SharglutDev) [inline front/tests/006-stdcm.spec.ts]: This test does almost the same thing as the 3rd expect checking the map and reset form. Can't both be merged ? or coudl'nt we create a global function to fill all inputs.

Or to add more custom behaviors, we could also check if the path layer is present on the simulation result map

File: front/src/applications/stdcm/components/StdcmHeader.tsx
Comment (SharglutDev) [inline front/src/applications/stdcm/components/StdcmHeader.tsx:28]: Do we need the `stdcm` keyword here ? We don't add `stdcm` before the other data-testid in these tests.

File: front/tests/005-operational-studies.spec.ts
Comment (SharglutDev) [inline front/tests/005-operational-studies.spec.ts:57]: Can't we use this new function also for 007 and 010 instead of `checkInfraLoaded` ?

File: front/tests/005-operational-studies.spec.ts
Comment (SharglutDev) [inline front/tests/005-operational-studies.spec.ts:22]: We need it because of the infra check ? Won't it be to slow for the CI ?

File: front/tests/pages/stdcm-page-model.ts
Comment (SharglutDev) [inline front/tests/pages/stdcm-page-model.ts:390]: Wouldn't we want to check if there is still a "addViaButton" before/after the new via ?

File: front/tests/006-stdcm.spec.ts
Comment (SharglutDev) [inline front/tests/006-stdcm.spec.ts:1]: Would it be useful to have a test where we check an O/D resulting to a pathinfing error to see if the warning box is there ?

File: front/tests/utils/api-setup.ts
Comment (SharglutDev) [inline front/tests/utils/api-setup.ts]: Do we have to use the syntax `{number}` for param/returns ? It's not displayed when we hover the function name.

File: front/tests/008-train-schedule.spec.ts
Comment (SharglutDev) [inline front/tests/008-train-schedule.spec.ts]: This a test on the scenario page, should we really use the stdcm datas ?

File: front/tests/008-train-schedule.spec.ts
Comment (SharglutDev) [inline front/tests/008-train-schedule.spec.ts:28]: Why are these values changing ? Should this really be part of this PR ?

File: front/tests/pages/stdcm-page-model.ts
Comment (SharglutDev) [inline front/tests/pages/stdcm-page-model.ts]:
```suggestion
    this.allViasButton = page.getByTestId('all-via-button');
```

File: front/tests/pages/stdcm-page-model.ts
Comment (SharglutDev) [inline front/tests/pages/stdcm-page-model.ts:297]: We don't test for the convoy ?

**Review by Akctarus — COMMENTED** [review]
Tested, I got: `Error: Infrastructure state did not reach 'CACHED' within the allotted 3 minutes.`

**Review by SharglutDev — COMMENTED** [review]
Thank you for this new PR, nice to see some tests for this new stdcm UI. Reviewed in peer with @Akctarus.

Left various comments.

The launch of the tests in debug mode went perfect at the first try and for the second try, the infra won't load and all tests are failing :/

**Comment by SharglutDev** [comment]
Also, since I didn't wait for tests to end, I'm stuck with small infra in stdcm when not in debug mode. How can I fix this ?

File: front/src/applications/stdcm/components/StdcmHeader.tsx
Comment (Maymanaf) [reply front/src/applications/stdcm/components/StdcmHeader.tsx:28]: I used the same structure as the class to be more consistent.

File: front/tests/005-operational-studies.spec.ts
Comment (Maymanaf) [reply front/tests/005-operational-studies.spec.ts:22]: The configured timeout for the tests is 90seconds, I added the slow so we can reach the 3 minutes of infra check.

File: front/tests/008-train-schedule.spec.ts
Comment (Maymanaf) [reply front/tests/008-train-schedule.spec.ts:28]: I updated the train schedule and the infrastructure used in this test, as well as in the STDCM test, so it's expected that the values have changed.

File: front/tests/pages/stdcm-page-model.ts
Comment (Maymanaf) [reply front/tests/pages/stdcm-page-model.ts:297]: If you mean the towed rolling stock, it will be added later when the related bugs are fixed #9544 and #9637

File: front/tests/006-stdcm.spec.ts
Comment (Maymanaf) [reply front/tests/006-stdcm.spec.ts:1]: Both a test for pathfinding errors and one for "no simulation found" could be useful, but they are not included in the scope of this ticket. I will consider addressing them later when adding towed rolling stocks and related paths.

File: front/tests/008-train-schedule.spec.ts
Comment (Maymanaf) [reply front/tests/008-train-schedule.spec.ts]: The same data is used for both tests to optimize test time, as loading the train schedule takes significant time.
I updated the names to "trainSchedule"

File: front/tests/pages/stdcm-page-model.ts
Comment (Maymanaf) [reply front/tests/pages/stdcm-page-model.ts:490]: Removed the verification in `downloadSimulation`

Comment (Maymanaf) [reply front/tests/pages/stdcm-page-model.ts:490]: Removed the check in the downloadSimulation function

File: front/tests/005-operational-studies.spec.ts
Comment (Maymanaf) [reply front/tests/005-operational-studies.spec.ts:57]: Done

File: front/tests/006-stdcm.spec.ts
Comment (Maymanaf) [reply front/tests/006-stdcm.spec.ts]: I combined both tests and I added a check for the path layer in the same test.

File: front/tests/pages/stdcm-page-model.ts
Comment (Maymanaf) [reply front/tests/pages/stdcm-page-model.ts]: Removed the click on the modal and added a comment to include the click on the close button once the feature is implemented.

File: front/tests/pages/stdcm-page-model.ts
Comment (Maymanaf) [reply front/tests/pages/stdcm-page-model.ts:390]: Done

File: front/tests/utils/api-setup.ts
Comment (Maymanaf) [reply front/tests/utils/api-setup.ts]: Removed the syntax for the different function.

**Comment by Maymanaf** [comment]
> Also, since I didn't wait for tests to end, I'm stuck with small infra in stdcm when not in debug mode. How can I fix this ?

@SharglutDev you should update stdcm env using `POST` `/api/stdcm/search_environment`

**Comment by Maymanaf** [comment]
> Tested, I got: `Error: Infrastructure state did not reach 'CACHED' within the allotted 3 minutes.`

@Akctars It's an `editoast `related problem; you should restart the Docker containers.

File: front/tests/global-teardown.ts
Comment (SharglutDev) [inline front/tests/global-teardown.ts:1]: Regarding the previous comment with the stdcm search env, maybe we should set it back to infra France after the tests ?

Comment (SharglutDev) [reply front/tests/global-teardown.ts:1]: More than infra France, we should set it back like it was before.

Comment (Maymanaf) [reply front/tests/global-teardown.ts:1]: I updated the code to restore the STDCM environment before running the tests locally, as there is no STDCM environment configured by default in the CI

**Comment by SharglutDev** [comment]
Lgtm and tested for me. Great job !

As @Yohh doesn't have his e2e maintainer privilege yet, I'm waiting for his approve to do mine.

File: front/src/assets/rollingStock/freightRollingStocks.ts
Comment (Yohh) [inline front/src/assets/rollingStock/freightRollingStocks.ts:115]: those rollingstocks are only used for tests, can't we declare them in an asset file in `tests/` then spread them in this array only if the tests are started?

Comment (Yohh) [inline front/src/assets/rollingStock/freightRollingStocks.ts:115]: those rollingstocks are only used for tests

in order to separate the usecases, can't we declare them in an asset file in `tests/` then spread them in this array only if the tests are started?

we could use `fs` at global setup to create the array to spread, this array will be empty if the tests are not lauched

Comment (Maymanaf) [reply front/src/assets/rollingStock/freightRollingStocks.ts:115]: The list of rolling stocks in `freightRollingStocks.ts` is a temporary solution for the project, not just for the tests. I believe we can leave it as is until the underlying logic is updated, at which point we can adjust the tests accordingly

Comment (Yohh) [reply front/src/assets/rollingStock/freightRollingStocks.ts:115]: ok, a comment would have been good to explain it, but it's not related to your issue

**Review by Yohh — APPROVED** [review]
lgtm

**Review by SharglutDev — APPROVED** [review]

**Review by Akctarus — APPROVED** [review]
Tested, LGTM, ty for this PR !

**Pull request merged**
